keras2onnx/ke2onnx/bidirectional.py

```diff
@@ -29,313 +24,14 @@ def _calculate_keras_bidirectional_output_shapes(operator):
-    lstm__type = 'LSTM'
-    lstm_input_names = []
-    lstm_output_names = []
-    lstm_attrs = {}
-
-    # Reshape Keras input format into ONNX input format
-    lstm_x_name = scope.get_unique_variable_name(operator.full_name + '_X')
-    apply_transpose(scope, operator.inputs[0].full_name, lstm_x_name, container, perm=[1, 0, 2])
-    lstm_input_names.append(lstm_x_name)
-
-    # Allocate input transformation matrix in ONNX and add its name into LSTM input list
-    tensor_w_name = scope.get_unique_variable_name(operator.full_name + '_W')
-    container.add_initializer(tensor_w_name, onnx_proto.TensorProto.FLOAT,
-                              [2, 4 * hidden_size, input_size], np.concatenate([W_x, W_x_back]).flatten())
-    lstm_input_names.append(tensor_w_name)
-
-    # Allocate hidden transformation matrix in ONNX and add its name into LSTM input list
-    tensor_r_name = scope.get_unique_variable_name(operator.full_name + '_R')
-    container.add_initializer(tensor_r_name, onnx_proto.TensorProto.FLOAT,
-                              [2, 4 * hidden_size, hidden_size], np.concatenate([W_h, W_h_back]).flatten())
-    lstm_input_names.append(tensor_r_name)
-
-    # Add bias vectors at different places in the original LSTM if needed
-    if b is not None:
-        tensor_b_name = scope.get_unique_variable_name(operator.full_name + '_B')
-        container.add_initializer(tensor_b_name, onnx_proto.TensorProto.FLOAT, [2, 8 * hidden_size],
-                                  np.concatenate([b, b_back]).flatten())
-        lstm_input_names.append(tensor_b_name)
-    else:
-        lstm_input_names.append('')  # the name of a non-existing optional variable is an empty string
-
-    # sequence_lens, this input is not used when converting Keras Bidirectional.
-    uses_masking_layer = len(operator.input_masks) == 1
-    if uses_masking_layer:
-        # Mask using sequence_lens input
-        sequence_lengths = scope.get_unique_variable_name(operator.full_name + '_seq_lens')
-        lstm_input_names.append(sequence_lengths)
-    else:
-        lstm_input_names.append('')
-    oopb = OnnxOperatorBuilder(container, scope)
-
-    if container.target_opset < 9:
-        # need the zero initializer to correct some engine shape inference bug.
-        state_shape = (2, 1, hidden_size)
-        initial_h_name = scope.get_unique_variable_name(operator.full_name + '_initial_h')
-        container.add_initializer(initial_h_name, onnx_proto.TensorProto.FLOAT, state_shape,
-                                  np.zeros(shape=state_shape).flatten())
-        lstm_input_names.append(initial_h_name)
-        initial_c_name = scope.get_unique_variable_name(operator.full_name + '_initial_c')
-        container.add_initializer(initial_c_name, onnx_proto.TensorProto.FLOAT, state_shape,
-                                  np.zeros(shape=state_shape).flatten())
-        lstm_input_names.append(initial_c_name)
+    op_type = operator.raw_operator.forward_layer.__class__
```

Comment:
Should  it be 'type(operator.raw_operator.forward_layer)', \__class__ seems a little python private thing.

Reply:
Agreed. :)